src/exchanges/populate.ts

```diff
@@ -0,0 +1,331 @@
+              kind: 'Name',
+              value: type,
+            },
+          },
+          name: {
+            kind: 'Name',
+            value: `${type}_PopulateFragment_${current.length}`,
+          },
+          selectionSet: selections,
+        },
+        type,
+      };
+      return {
+        ...state,
+        [type]: [...current, entry],
+      };
+    }, typeFragments);
+  };
+
+  const handleIncomingTeardown = ({ key, operationName }: Operation) => {
+    if (operationName !== 'teardown') {
+      return;
+    }
+
+    activeOperations.delete(key);
+  };
+
+  return ops$ => {
+    return pipe(
+      ops$,
+      tap(handleIncomingQuery),
+      tap(handleIncomingTeardown),
+      map(handleIncomingMutation),
+      forward
+    );
+  };
+};
+
+type UserFragmentMap<T extends string = string> = Record<
+  T,
+  FragmentDefinitionNode
+>;
+
+type TypeFragmentMap<T extends string = string> = Record<T, TypeFragment[]>;
+
+interface TypeFragment {
+  /** Operation key where selection set is being used. */
+  key: number;
+  /** Selection set. */
+  fragment: FragmentDefinitionNode;
+  /** Type of selection. */
+  type: string;
+}
+
+interface MakeFragmentsFromQueryArg {
+  schema: GraphQLSchema;
+  query: DocumentNode;
+}
+
+/** Gets typed selection sets and fragments from query */
+export const extractSelectionsFromQuery = ({
+  schema,
+  query,
+}: MakeFragmentsFromQueryArg) => {
+  let selections: { selections: SelectionSetNode; type: string }[] = [];
+  let fragments: FragmentDefinitionNode[] = [];
+  const typeInfo = new TypeInfo(schema);
+
+  visit(
+    query,
+    visitWithTypeInfo(typeInfo, {
+      Field: node => {
+        if (!node.selectionSet) {
+          return undefined;
+        }
+
+        const type = getType(typeInfo);
+
+        if (!type) {
+          return undefined;
+        }
+
+        selections = [...selections, { selections: node.selectionSet, type }];
+      },
+      FragmentDefinition: node => {
+        fragments = [...fragments, node];
+      },
+    })
+  );
+
+  return { selections, fragments };
+};
+
+interface AddFragmentsToQuery {
+  schema: GraphQLSchema;
+  query: DocumentNode;
+  typeFragments: Record<string, Omit<TypeFragment, 'key'>[]>;
+  userFragments: UserFragmentMap;
+}
+
+/** Replaces populate decorator with fragment spreads + fragments. */
+export const addFragmentsToQuery = ({
+  schema,
+  query,
+  typeFragments,
+  userFragments,
+}: AddFragmentsToQuery) => {
+  const typeInfo = new TypeInfo(schema);
+  const requiredUserFragments = new Set<FragmentDefinitionNode>();
+  let additionalFragments: Record<string, FragmentDefinitionNode> = {};
+
+  return visit(
+    query,
+    visitWithTypeInfo(typeInfo, {
+      Field: {
+        enter: node => {
+          if (
+            !node.directives ||
+            !node.directives.find(d => d.name.value === 'populate')
+          ) {
+            return;
+          }
+
+          const directives = node.directives.filter(
```

Comment:
nit: you could avoid the double iteration (find + filter) by checking `!node.directives` once and checking `directives.length < node.directives.length` below instead 👀

Reply:
For real, we need to be prepared for nodes which have 1000s+ directives! 😉 

But more seriously, this is probably a good time to mention that there are some potential performance improvements that can be done (e.g. `getUsedFragments` which calls `visit` inside of an existing `visit` (multiple traversals of the same subtree).

If my thinking is correct, this would be a time complexity of `O(2n)` which should be negligible but I'd like to see this in practice on larger structures before resting on that assumption.

Reply:
We could always optimise this in a separate PR though, so let’s maybe not worry about that yet? All the visiting logic in Graphcache is custom. The AST is pretty compact and easy to traverse so we could also make some specialised traversal logic if that makes things easier

Reply:
Yeah lets visit that in a later PR.

Would be great to try this exchange out on some larger ASTs